src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -31,11 +31,113 @@ public struct CallGraphNode
             public QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
         }
 
+        // TODO: 
+        // This is the method that should be invoked to verify cycles of interest,
+        // i.e. where each callable in the cycle is type parametrized.
+        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined. 
+        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        {
+            var parent = rootNode.CallableName;
+            bool EqualsParent(QsQualifiedName origin) => origin.Namespace.Value == parent.Namespace.Value && origin.Name.Value == parent.Name.Value;
+            var validResolution = TryCombineTypeResolutions(parent, out var combined, edges.Select(edge => edge.ParamResolutions).ToArray());
+            var resolvedToConcrete = combined.Values.All(res => !(res.Resolution is ResolvedTypeKind.TypeParameter tp) || EqualsParent(tp.Item.Origin));
```

Comment:
The parent's type parameter is considered to be concrete within a parent. Also, this will return true for resolutions to concrete types as well, so resolvedToParent would not be accurate.